From 3ba2df874ca27f075a0897210d8dce6c56646e60 Mon Sep 17 00:00:00 2001
From: Bartosz Golaszewski <bartekgola@gmail.com>
Date: Fri, 07 Feb 2014 16:14:37 +0000
Subject: grep: properly handle grep -w "^str" in !EXTRA_COMPAT case too

function                                             old     new   delta
grep_file                                           1267    1288     +21

Signed-off-by: Bartosz Golaszewski <bartekgola@gmail.com>
Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
(limited to 'findutils/grep.c')

diff --git a/findutils/grep.c b/findutils/grep.c
index a7bc4ca..7685946 100644
--- a/findutils/grep.c
+++ b/findutils/grep.c
@@ -375,6 +375,8 @@ static int grep_file(FILE *file)
 			} else {
 #if ENABLE_EXTRA_COMPAT
 				unsigned start_pos;
+#else
+				int match_flg;
 #endif
 				char *match_at;
 
@@ -392,6 +394,7 @@ static int grep_file(FILE *file)
 #if !ENABLE_EXTRA_COMPAT
 				gl->matched_range.rm_so = 0;
 				gl->matched_range.rm_eo = 0;
+				match_flg = 0;
 #else
 				start_pos = 0;
 #endif
@@ -400,7 +403,7 @@ static int grep_file(FILE *file)
 //bb_error_msg("'%s' start_pos:%d line_len:%d", match_at, start_pos, line_len);
 				if (
 #if !ENABLE_EXTRA_COMPAT
-					regexec(&gl->compiled_regex, match_at, 1, &gl->matched_range, 0) == 0
+					regexec(&gl->compiled_regex, match_at, 1, &gl->matched_range, match_flg) == 0
 #else
 					re_search(&gl->compiled_regex, match_at, line_len,
 							start_pos, /*range:*/ line_len,
@@ -415,7 +418,7 @@ static int grep_file(FILE *file)
 						found = 1;
 					} else {
 						char c = ' ';
-						if (gl->matched_range.rm_so)
+						if (match_at > line || gl->matched_range.rm_so != 0)
 							c = match_at[gl->matched_range.rm_so - 1];
 						if (!isalnum(c) && c != '_') {
 							c = match_at[gl->matched_range.rm_eo];
@@ -432,6 +435,7 @@ static int grep_file(FILE *file)
 #if !ENABLE_EXTRA_COMPAT
 								if (gl->matched_range.rm_eo != 0) {
 									match_at += gl->matched_range.rm_eo;
+									match_flg |= REG_NOTBOL;
 									goto opt_w_again;
 								}
 #else
--
cgit v0.9.1
